Update EIP-6206: clarify wording on JUMPF (EOF) (#9055)
ethereum/EIPs: EIPS/eip-6206.md

  * either `type[current_section_index].outputs` MUST be greater or equal `type[target_section_index].outputs`,
  * or `type[target_section_index].outputs` MUST be `0x80`
* The stack height validation at `JUMPF` depends on whether the target section is non-returning:

>   * `JUMPF` into returning section (`type[target_section_index].outputs` does not equal `0x80`):  `stack_height_min` and `stack_height_max` MUST be equal to `type[current_section_index].outputs + type[target_section_index].inputs  - type[target_section_index].outputs`. This means that target section can output less stack elements than the original code section called by the top element on the return stack, if the current code section leaves the delta `type[current_section_index].outputs - type[target_section_index].outputs` element(s) on the stack.

Author: Heh, OK, I think you're right again... having spent some time with the rationale section, I'm starting to see it's maybe rationale section repeating the spec, and this sentence is ok to stay here as clarification...

* Stack overflow check at `JUMPF`: `stack_height_max` MUST be less than or equal to `1024 - types[target_section_index].max_stack_height + types[target_section_index].inputs`.
* `JUMPF` is considered terminating instruction, i.e. does not have successor instructions in code validation and MAY be final instruction in the section. 
* The code validation defined in [EIP-4200](./eip-4200.md) also fails if any `RJUMP*` offset points to one of the two bytes directly following a `JUMPF` instruction.

`CALLF` instruction validation is extended to include the rule:

* Code section is invalid in case an immediate argument `target_section_index` of any `CALLF` targets a non-returning section, i.e. `type[target_section_index].outputs` equals `0x80`.

#### Non-returning status validation

Section type MUST be non-returning if and only if the section contains no `RETF` instructions and no `JUMPF` instructions targeting returning sections (target section's status is checked via its output value in type section.)
*Note: This implies that section containing only `JUMPF`s into non-returning sections is non-returning itself.*

## Rationale

### Allowing `JUMPF` to section with less outputs

> A code section with a `JUMPF` can, just before that instruction, have a delta of `type[current_section_index].outputs - type[target_section_index].outputs` output elements on the operand stack (i.e. counted on top of the target section inputs!). This makes it possible to `JUMPF` to a section with less outputs than those of the current section. This will reduce duplicated code as it will allow compilers more flexibility during code generation such that certain helpers can be used generically by functions, regardless of their output values.

Reviewer: > (i.e. counted on top of the target section inputs!).

This is quite confusing, output elements are _below_ target section inputs, right? I would just remove the part in parenthesis.

> A code section with a `JUMPF` can, just before that instruction, have a delta of `type[current_section_index].outputs - type[target_section_index].outputs` output elements on the operand stack (i.e. counted on top of the target section inputs!). This makes it possible to `JUMPF` to a section with less outputs than those of the current section. This will reduce duplicated code as it will allow compilers more flexibility during code generation such that certain helpers can be used generically by functions, regardless of their output values.

Reviewer: ```suggestion
As long as code section with `JUMPF` has the delta `type[current_section_index].outputs - type[target_section_index].outputs` stack elements before changing code sections, it is possible to `JUMPF` to a section with less outputs than those of the current section. This will reduce duplicated code as it will allow compilers more flexibility during code generation such that certain helpers can be used generically by functions, regardless of their output values.
```

Author: ok @gumb0 PTAL now. I brought back the sentence in the spec you've requested, but gave the rationale piece a less redundant sound and hopefully more to the point of why we're having this.

* Stack overflow check at `JUMPF`: `stack_height_max` MUST be less than or equal to `1024 - types[target_section_index].max_stack_height + types[target_section_index].inputs`.
* `JUMPF` is considered terminating instruction, i.e. does not have successor instructions in code validation and MAY be final instruction in the section. 
* The code validation defined in [EIP-4200](./eip-4200.md) also fails if any `RJUMP*` offset points to one of the two bytes directly following a `JUMPF` instruction.

`CALLF` instruction validation is extended to include the rule:

* Code section is invalid in case an immediate argument `target_section_index` of any `CALLF` targets a non-returning section, i.e. `type[target_section_index].outputs` equals `0x80`.

#### Non-returning status validation

Section type MUST be non-returning if and only if the section contains no `RETF` instructions and no `JUMPF` instructions targeting returning sections (target section's status is checked via its output value in type section.)
*Note: This implies that section containing only `JUMPF`s into non-returning sections is non-returning itself.*

## Rationale

### Allowing `JUMPF` to section with less outputs

> An alternative rule for `JUMPF` stack validation could require the target section's outputs to be exactly equal to the current section's outputs. Under such rule, a particular target section (a shared "helper" piece of code) would only "match" sections (requiring some shared "helper" code to execute before returning) *of the same outputs*.

Reviewer: ```suggestion
An alternative rule for `JUMPF` stack validation could require the target section's outputs to be exactly equal to the current section's outputs. Under such rule, a particular target section (a shared "helper" piece of code) would only "match" sections (requiring some shared "helper" code to execute before returning) *with the same number of outputs*.
```

not sure actually if the second sentence here adds anything to understanding

Author: it is a bit redundant, but describing the exact problem with the alternative helps (at least me) to follow the reasoning more easily - it introduces the second sentence of the (now) second paragraph about code gen. It also speaks to the concept of helpers from the Motivation section, bridging it to the gist of the "delta spec".

So the narrative of the 4 sentences is:
- what is the alternative spec
- why is it bad
- what is the chosen spec
- why is it good

Reviewer 2: It does sound clearer to me so I think it's an improvement.
